File: Python solutions/Strings/valid_palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:

        #  We are using 2 pointer approach so initialize 
        # left pointer to 0 and right pointer to str length - 1.
        
        left = 0
        right = len(s) - 1

        # Main idea is to start from extreme ends and keep moving inwards till 
        # we don't get valid (alphanumeric) characters at both pointer indexes.

        # We check whether these two valid values are equal, if yes we move both
        # pointers inward by 1. Else we return false. 

        while left < right:

            while (left < right) and (not self.isAlphaNumeric(s[left])):
                left += 1

            while (left < right) and (not self.isAlphaNumeric(s[right])):
                right -= 1

            # We convert both valid character to lowercase as case doesn't matter in this problem.

            leftChar = s[left].lower()
            rightChar = s[right].lower()

            if leftChar != rightChar:
                return False
            else:
                left += 1
                right -= 1

        return True

# A function which checks whether a character is alphanumeric or not.

    def isAlphaNumeric(self, c: str) -> bool:
        return (c >= 'A' and c <= 'Z') or (c >= 'a' and c <= 'z') or (c >= '0' and c <= '9')

File: Java solutions/Strings/validPalindrome.java
class Solution {
    public boolean isPalindrome(String s) {

        // We are using 2 pointer approach so initialize 
        // left pointer to 0 and right pointer to str length - 1.
        
        int left = 0;
        int right = s.length() - 1;

        // Main idea is to start from extreme ends and keep moving inwards till 
        // we don't get valid (alphanumeric) characters at both pointer indexes.

        // We check whether these two valid values are equal, if yes we move both
        // pointers inward by 1. Else we return false. 

        while(left < right)
        {
            while(left < right && !isAlphaNumeric(s.charAt(left)))
            {
                left++;
            }

            while(left < right && !isAlphaNumeric(s.charAt(right)))
            {
                right--;
            }

            // We convert both valid character to lowercase as case doesn't matter in this problem.

            char leftChar = Character.toLowerCase(s.charAt(left));
            char rightChar = Character.toLowerCase(s.charAt(right));

            if(leftChar != rightChar)
            {
                return false;
            }
            else
            {
                left++;
                right--;
            }
        }

        return true;
        
    }

// A function which checks whether a character is alphanumeric or not.

    private boolean isAlphaNumeric(char c)
    {
        return (c >= 'A' && c <= 'Z') ||
                 (c >= 'a' && c <= 'z') ||
                 (c >= '0' && c <= '9');
    }
}